focus-android/app/src/main/java/org/mozilla/focus/shortcut/HomeScreen.kt

```diff
@@ -0,0 +1,122 @@
+
+    /**
+     * Create a shortcut for the given website on the device's home screen.
+     */
+    @Suppress("LongParameterList")
+    fun installShortCut(
+        context: Context,
+        icon: Bitmap,
+        url: String?,
+        title: String,
+        blockingEnabled: Boolean,
+        requestDesktop: Boolean,
+    ) {
+        val shortcutTitle = if (TextUtils.isEmpty(title.trim { it <= ' ' })) {
+            generateTitleFromUrl(url)
+        } else {
+            title
+        }
+
+        installShortCutViaManager(context, icon, url, shortcutTitle, blockingEnabled, requestDesktop)
+
+        // Creating shortcut flow is different on Android up to 7, so we want to go
+        // to the home screen manually where the user will see the new shortcut appear
+        if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.N_MR1) {
+            goToHomeScreen(context)
+        }
+    }
+
+    /**
+     * Create a shortcut via the [ShortcutManagerCompat].
+     *
+     * On Android versions up to 7 shortcut will be created via system broadcast internally.
+     *
+     * On Android 8+ the user will have the ability to add the shortcut manually
+     * or let the system place it automatically.
+     */
+    @Suppress("LongParameterList")
+    private fun installShortCutViaManager(
+        context: Context,
+        bitmap: Bitmap,
+        url: String?,
+        title: String,
+        blockingEnabled: Boolean,
+        requestDesktop: Boolean,
+    ) {
+        if (ShortcutManagerCompat.isRequestPinShortcutSupported(context)) {
+            val icon = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+                IconCompat.createWithAdaptiveBitmap(bitmap)
+            } else {
+                IconCompat.createWithBitmap(bitmap)
+            }
+            val shortcut = ShortcutInfoCompat.Builder(context, UUID.randomUUID().toString())
+                .setShortLabel(title)
+                .setLongLabel(title)
+                .setIcon(icon)
+                .setIntent(createShortcutIntent(context, url, blockingEnabled, requestDesktop))
+                .build()
+            ShortcutManagerCompat.requestPinShortcut(context, shortcut, null)
+        }
+    }
+
+    private fun createShortcutIntent(
+        context: Context,
+        url: String?,
+        blockingEnabled: Boolean,
+        requestDesktop: Boolean,
+    ): Intent {
+        val shortcutIntent = Intent(context, MainActivity::class.java)
+        shortcutIntent.action = Intent.ACTION_VIEW
+        shortcutIntent.data = Uri.parse(url)
+        shortcutIntent.putExtra(BLOCKING_ENABLED, blockingEnabled)
+        shortcutIntent.putExtra(REQUEST_DESKTOP, requestDesktop)
+        shortcutIntent.putExtra(ADD_TO_HOMESCREEN_TAG, ADD_TO_HOMESCREEN_TAG)
+        return shortcutIntent
+    }
+
+    /**
+     * Generates a new default title based on the URL.
+     */
+    @VisibleForTesting
+    fun generateTitleFromUrl(url: String?): String {
```

Comment:
I don't think we should allow a nullable String here, wdyt? Should we also handle the case when 'host' is null

Reply:
Good call! A null string would throw NPE.